bookShelf/books/views.py

```diff
@@ -17,147 +19,422 @@
+			for k in range(len(tags)):
+				if tags[k] not in keys:
+					metafile.write(tags[k] + '\n')
+			metafile.close()
+		return render(request, 'books/thanks.html')
 	else:
-		profs = json.loads(open("profs.json","r").read(), object_pairs_hook=OrderedDict)
-		return render(request, 'books/upload.html',{"profs":profs})
+		profs = json.loads(open("profs.json", "r").read(), object_pairs_hook=OrderedDict)
+		return render(request, 'books/upload.html', {"profs": profs})
+
+
+def download_course(request):
+	"""
+		function to serve the zip files of entire courses
+	"""
+	course = request.GET.get('course', 'none')
+	if course == 'none':
+		return redirect('/books/')
+	parent_dir = course[0:2]
+	zip_location = os.path.join(DATABASE_DIR, parent_dir, course + '.zip')
+	if not (os.path.exists(zip_location) and os.path.isfile(zip_location)):
+		course_path = os.path.join(DATABASE_DIR, parent_dir, course)
+		zf = zipfile.ZipFile(zip_location, "w")
+		for dirname, sudirs, files in os.walk(course_path):
+			for filename in files:
+				if not filename.lower().endswith('.meta'):
+					path = os.path.join(dirname, filename)
+					to_write = os.path.relpath(path, course_path)
+					zf.write(path, arcname=to_write)
+		zf.close()
+	"""
+		records the downloaded zip file in the stats file
+	"""
+	with open(STATS_FILE, "r") as file:
+		stats = json.load(file)
+	stats[course]["downloads"] = stats[course]["downloads"] + 1
+	stats[course]["last"] = datetime.now().strftime("%d/%m/%Y")
+	with open(STATS_FILE, "w") as file:
+		json.dump(stats, file)
+	loc = DATABASE_DIR.split(os.sep)
+	path = '/'.join(loc)
+	return redirect('/' + path + '/' + parent_dir + '/' + course + '.zip')
+
 
-## Controller to Handle approval of requests
 @login_required
-def approve(request):	
-	unapproved_documents = []
+def approve(request):
+	"""
+		Controller to Handle approval of requests
+	"""
+	unapproved_documents = {}
 	for path, subdirs, files in os.walk(UNAPPROVED_DIR):
 		for filename in files:
-			f = os.path.join(path, filename)
-			unapproved_documents.append(str(f)[str(f).rindex("/")+1:])
-	if(len(unapproved_documents)==0):
+			if not filename.endswith('.meta'):
+				arg = False
+				f = os.path.join(path, filename)
+				name = f.split(SEPARATOR)[-1]
+				check = os.path.join(BULK_UP_DIR, name)
+				print(check)
+				if os.path.exists(check) and os.path.isfile(check):
+					arg = True
+				unapproved_documents[str(f)[str(f).rindex(os.sep) + 1:]] = arg
+
+	if len(unapproved_documents) == 0:
 		error = "No Unapproved documents present, please ask people to upload material and contribute to the Citadel"
 	else:
-		error=''
-	return render(request, 'books/approve.html', {'unapproved_documents':unapproved_documents,'error': error})
+		error = ''
+	return render(request, 'books/approve.html', {'unapproved_documents': unapproved_documents, 'error': error})
+
 
 @login_required
 def remove_unapproved_document(request):
+	name = request.GET.get('name', 'none')
+	metafile = name + '.meta'
 	try:
```

Comment:
Do this in 2 separate try catch blocks. It is possible that first one causes error and so .meta file will bot be deleted